tests/tests_fabric/utilities/test_optimizer.py

```diff
@@ -1,36 +1,131 @@
-        if isinstance(param, Tensor):
-            assert param.data.device.type == device
+    class CachedRandomTensorDataset(torch.utils.data.Dataset):
+        """Very low overhead torch dataset for training for a given number of steps."""
+
+        def __init__(self, batch_size: int, num_features: int, num_responses: int, length: int, device: str) -> None:
+            self.x = torch.randn((batch_size, num_features), device=torch.device(device))
+            self.y = torch.randn((batch_size, num_responses), device=torch.device(device))
+            self.length = length
+
+        def __getitem__(self, idx: int) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+            return self.x.clone(), self.y.clone()
+
+        def __len__(self) -> int:
+            return self.length
+
+    def simple_training(optimizer, model, dataset, loss_fn):
+        for input, target in dataset:
+            optimizer.zero_grad()
+            output = model(input)
+            loss = loss_fn(output, target)
+            loss.backward()
+            optimizer.step()
+
+    if not torch.cuda.is_available():
+        return
+
+    gpu_device = "cuda"
+    devices = ["cpu", gpu_device]
+
+    num_features = 32
+    num_responses = 2
+    batch_size = 16
+
+    optimizer_on_device = {}
+    for device in devices:
+        dataset = CachedRandomTensorDataset(batch_size, num_features, num_responses, batch_size * 16, device)
+        dataloader = DataLoader(dataset, batch_size=None, shuffle=False)
+        model = torch.nn.Linear(num_features, num_responses)
+        model.to(device=device)
+        fused_vals = [False] if device == "cpu" else [False, True]
+
+        for fused in fused_vals:
+            optimizer = torch.optim.Adam(model.parameters(), lr=0.1, fused=fused)
+            simple_training(optimizer, model, dataloader, loss_fn=nn.MSELoss())
+            if fused:
+                optimizer_on_device[device + "_fused_" + str(fused)] = optimizer
+            else:
+                optimizer_on_device[device] = optimizer
+
+    # Test _optimizer_to_device function
+    # Test cpu-->gpu, fused = False from CPU
+    opt_to_gpu = copy.deepcopy(optimizer_on_device["cpu"])
+    _optimizer_to_device(opt_to_gpu, gpu_device)
+    assert_opt_state_in_expected_location(opt_to_gpu, optimizer_on_device[gpu_device])
+
+    # Test gpu-->cpu, fused = False
+    opt_to_cpu = copy.deepcopy(optimizer_on_device[gpu_device])
+    _optimizer_to_device(opt_to_cpu, "cpu")
+    assert_opt_state_in_expected_location(opt_to_cpu, optimizer_on_device["cpu"])
+
+    # Test gpu-->cpu, fused = True
+    opt_to_cpu = copy.deepcopy(optimizer_on_device[gpu_device + "_fused_True"])
+    _optimizer_to_device(opt_to_cpu, "cpu")
+    assert_opt_state_in_expected_location(opt_to_cpu, optimizer_on_device["cpu"])
+
+    # Try from_dict
+    # These all pretend that we have an appropriate prototype, I don't think we can actually do this since
+    # all we may have is a CPU pickle
```

Comment:
The test we have for pytorch `load_state_dict` being able to read a CPU checkpoint into an appropriate GPU optimizer is here: https://github.com/pytorch/pytorch/blob/main/test/test_optim.py#L1545-L1574

The code above is also how I expect checkpointing to happen, without the need of an explicit move to device.

Reply:
I have updated this test to be more explicit about what is going on, please take a look and see if it makes sense since the test you linked doesn't look at thorough as far as I can tell.

Reply:
Yea, the case we test is moving from CPU to GPU, and I see you test more combinations.